docs/proposals/blocks-storage-sharding.md

```diff
@@ -0,0 +1,77 @@
+
+1. **Periodically**<br />
+   to discover new blocks uploaded by ingesters or compactor, and delete old blocks removed due to retention or by the compactor
+2. **On-demand**<br/>
+   when the ring topology changes (the tokens ranges assigned to the gateway instance have changed)
+
+It's important to outline that the sync takes time (typically will have to re-scan the bucket and download new blocks index headers) and Cortex needs to guarantee query results consistency at any given time (_see below_).
+
+### Query execution
+
+When a querier executes a query, it will need to fetch series both from ingesters and the store-gateway instances.
+
+For a given query, the number of blocks to query is expected to be low, especially if the Cortex cluster is running the `query-frontend` with a `24h` query split interval. In this scenario, whatever is the client's query time range, the `query-frontend` will split the client's query into sub-queries each with up to `24h` time range and the querier will likely hit not more than 1 block per sub-query (except for the last 24h for which blocks may have not been compacted yet).
+
+Given this assumption, we want to avoid sending every query to every store-gateway instance. The querier should be able to take an informed decision about the minimum subset of store-gateway instances which needs to query given a time range.
+
+The idea is to run the [`MetaFetcher`](https://github.com/thanos-io/thanos/blob/master/pkg/block/fetcher.go#L127) also within the querier, but without any sharding filter (contrary to the store-gateway). At any given point in time, the querier knows the entire list of blocks in the storage. When the querier executes the `Select()` (or `SelectSorted()`) it does:
+
+1. Compute the list of blocks by the query time range
+2. Compute the minimum list of store-gateway instances containing the required blocks (using the information from the ring)
+3. Fetch series from ingesters and the matching store-gateway instances
+4. Merge and deduplicate received series
+   - Optimization: can be skipped if the querier hits only 1 store-gateway
+
+### Query results consistency
+
+When a querier executes a query, it should guarantee that either all blocks matching the time range are queried or the query fails.
+
+However, due to the (intentional) lack of a strong coordination between queriers and store-gateways, and the ring topology which can change any time, there's no guarantee that the blocks assigned to a store-gateway shard are effectively loaded on the store-gateway itself at any given point in time.
+
+The idea is introduce a **consistency check in the querier**. When a store-gateway receives a request from the querier, the store-gateway includes in the response the list of block IDs currently loaded on the store-gateway itself. The querier can then merge the list of block IDs received from all store-gateway hit, and match it against the list of block IDs computed at the beginning of the query execution.
```

Comment:
> the store-gateway includes in the response the list of block IDs currently loaded on the store-gateway itself

Isn't this going to be big? Could store-gateway return block IDs that were used to handle the query?

Alternatively, could querier tell store-gateway, which block it expects that store-gateway should use? (List of blocks it "must" use, and list of blocks it "could" use -- for new blocks. This can also trigger store-gateway to do a new scan for those blocks, if they are missing).

Reply:
> Isn't this going to be big?

Given a query is bounded to 1 tenant, 1 block per day = 365 blocks per 1 year. Considering a block ID is 16 bytes, 365 blocks = 5.7KB.

> Could store-gateway return block IDs that were used to handle the query?

The store-gateway should just need to return the blocks matching the query time range. It's an optimization we can/should put in place.

Reply:
This is actually a good fit for `Info` (something we have in StoreAPI) endpoint rather... Have you considered this?

Reply:
With such Info endpoint you really don't need querier consistency check at all ,right? :thinking:

Reply:
Yep the problem is when do we query the info endpoint? Periodically, before we send the query, or after we send the query?

The problem with periodically and after we send the query is that the list is not consistent, but maybe we could do it right before we send the query to the -gateway. But do we really want to do 2 RPC calls when we could do one?

Reply:
periodically.

Reply:
if you do it before, after or even in the middle! There will be inconsistency **anyway** ... This how distributed systems work ;p

Reply:
Not if its in the same call ;)

Again the problem is we don't want to return a partial result. The problem is suppose, we're doing it every 10secs. And we synced at `t0`, and the topology changed at `t0+8s`, and we sent a query at `t0+9s`. We might have missed a block because the gateway doesn't have it anymore but the querier thinks the gateway has it and might respond with a gap in the result.

It is not OK in the case of Cortex to return inconsistent result. We'd rather fail the query.

Reply:
Good point, missed that. With the same call and addting that to response, can help to give more consistent insight :+1: 
Overhead does not problematic.